src/Processors/Formats/Impl/Parquet/keyConditionRPNToParquetRPN.cpp

```diff
@@ -197,202 +230,32 @@ std::optional<std::vector<uint64_t>> hash(const IColumn * data_column, const par
-
-    for (const auto & element : condition)
-    {
-        if (element.function == Function::FUNCTION_IN
-            || element.function == Function::FUNCTION_NOT_IN)
-        {
-            bool maybe_true = true;
-            for (auto column_index = 0u; column_index < element.hashes_per_column.size(); column_index++)
-            {
-                // in case bloom filter is not present for this row group
-                // https://github.com/ClickHouse/ClickHouse/pull/62966#discussion_r1722361237
-                if (!column_index_to_column_bf.contains(element.key_columns[column_index]))
-                {
-                    rpn_stack.emplace_back(true, true);
-                    continue;
-                }
-
-                bool column_maybe_contains = maybeTrueOnBloomFilter(
-                    element.hashes_per_column[column_index],
-                    column_index_to_column_bf.at(element.key_columns[column_index]));
-
-                if (!column_maybe_contains)
-                {
-                    maybe_true = false;
-                    break;
-                }
-            }
-
-            rpn_stack.emplace_back(maybe_true, true);
-            if (element.function == Function::FUNCTION_NOT_IN)
-                rpn_stack.back() = !rpn_stack.back();
-        }
-        else if (element.function == Function::FUNCTION_NOT)
-        {
-            rpn_stack.back() = !rpn_stack.back();
-        }
-        else if (element.function == Function::FUNCTION_OR)
-        {
-            auto arg1 = rpn_stack.back();
-            rpn_stack.pop_back();
-            auto arg2 = rpn_stack.back();
-            rpn_stack.back() = arg1 | arg2;
-        }
-        else if (element.function == Function::FUNCTION_AND)
-        {
-            auto arg1 = rpn_stack.back();
-            rpn_stack.pop_back();
-            auto arg2 = rpn_stack.back();
-            rpn_stack.back() = arg1 & arg2;
-        }
-        else if (element.function == Function::ALWAYS_TRUE)
-        {
-            rpn_stack.emplace_back(true, false);
-        }
-        else if (element.function == Function::ALWAYS_FALSE)
-        {
-            rpn_stack.emplace_back(false, true);
-        }
-        else
-        {
-            rpn_stack.emplace_back(true, true);
-        }
-    }
-
-    if (rpn_stack.size() != 1)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected stack size in KeyCondition::mayBeTrueOnRowGroup");
-
-    return rpn_stack[0].can_be_true;
-}
-
-std::unordered_set<std::size_t> ParquetBloomFilterCondition::getFilteringColumnKeys() const
+KeyCondition::RPN keyConditionRPNToParquetRPN(const std::vector<KeyCondition::RPNElement> & rpn,
+                                              const std::vector<ArrowFieldIndexUtil::ClickHouseIndexToParquetIndex> & clickhouse_column_index_to_parquet_index,
+                                              const std::unique_ptr<parquet::RowGroupMetaData> & parquet_rg_metadata)
```

Comment:
What I had in mind is for ParquetBlockInputFormat to make a copy of KeyCondition and modify it. So pretty much the same as what this PR does now, just cosmetically different (the modified KeyCondition::RPN + other fields would be wrapped in a KeyCondition instead of passed individually to checkRPNAgainstHyperrectangle).

Reply:
So.. something like:

`
auto parquet_key_condition = key_condition->clone();
parquet_key_condition->prepareBloomFilterData();
parquet_key_condition->checkInHyperrectangle(..., column_index_to_bf);
`

Just unclear about the clone or copy constructor of `KeyCondition`, but that might worrk